### PR TITLE
Client compression

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/grpc/grpc-dotnet/issues/232

### CODE REVIEW

**Comment by JamesNK** [comment]
@jtattermusch Please review that this is doing client compression correctly.

File: src/Grpc.Net.Client/Internal/Compression/GzipCompressionProvider.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
+﻿#region Copyright notice and license
```
Comment (JunTaoLuo) [inline src/Grpc.Net.Client/Internal/Compression/GzipCompressionProvider.cs]: This is the same as the GzipCompressionProvider in Grpc.AspNetCore.Server. Can we create a shared folder and add it to the Compile itemgroup instead?

File: src/Grpc.Net.Client/Internal/Compression/ICompressionProvider.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
+﻿#region Copyright notice and license
```
Comment (JunTaoLuo) [inline src/Grpc.Net.Client/Internal/Compression/ICompressionProvider.cs]: Same here

File: src/Grpc.Net.Client/Internal/Compression/GzipCompressionProvider.cs (hunk 1 above)
Comment (JamesNK) [reply src/Grpc.Net.Client/Internal/Compression/GzipCompressionProvider.cs]: This is internal, the server is not. I might make the server compression API internal, but unifying code will bloat this PR. Will do it separately.

File: testassets/InteropTestsClient/RunTests.ps1 [hunk 3]
```diff
@@ -23,10 +23,9 @@ $allTests =
   "status_code_and_message",
   "special_status_message",
   "unimplemented_service",
-  "unimplemented_method"
-  #,
-  #"client_compressed_unary",
-  #"client_compressed_streaming"
+  "unimplemented_method",
+  "client_compressed_unary",
```
Comment (jtattermusch) [inline testassets/InteropTestsClient/RunTests.ps1]: I assume these are passing now?

Comment (JamesNK) [reply testassets/InteropTestsClient/RunTests.ps1]: Yes

File: src/Grpc.Net.Client/Internal/GrpcCall.cs [hunk 4]
```diff
@@ -506,18 +517,26 @@ private HttpRequestMessage CreateHttpRequestMessage()
...
                     {
+                        // grpc-timeout is set via CallOptions.Deadline
                         continue;
                     }
-
-                    AddHeader(message.Headers, entry);
+                    else if (entry.Key == GrpcProtocolConstants.CompressionRequestAlgorithmHeader)
+                    {
+                        // grpc-internal-encoding-request is used in the client to set message compression
+                        message.Headers.Add(GrpcProtocolConstants.MessageEncodingHeader, entry.Value);
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Grpc.Net.Client/Internal/GrpcCall.cs]: That's what I thought.

When we have the concept of channel I'd like to add a strongly typed property for compression. I think it is fine to still have this special header then, it doesn't hurt anyone.

**Pull request merged**